handler/src/main/java/io/netty/handler/ssl/AbstractSniHandler.java

```diff
@@ -0,0 +1,323 @@
+    private static final int MAX_SSL_RECORDS = 4;
+
+    private static final InternalLogger logger =
+            InternalLoggerFactory.getInstance(AbstractSniHandler.class);
+
+    private boolean handshakeFailed;
+    private boolean suppressRead;
+    private boolean readPending;
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if (!suppressRead && !handshakeFailed) {
+            final int writerIndex = in.writerIndex();
+            try {
+                loop:
+                for (int i = 0; i < MAX_SSL_RECORDS; i++) {
+                    final int readerIndex = in.readerIndex();
+                    final int readableBytes = writerIndex - readerIndex;
+                    if (readableBytes < SslUtils.SSL_RECORD_HEADER_LENGTH) {
+                        // Not enough data to determine the record type and length.
+                        return;
+                    }
+
+                    final int command = in.getUnsignedByte(readerIndex);
+
+                    // tls, but not handshake command
+                    switch (command) {
+                        case SslUtils.SSL_CONTENT_TYPE_CHANGE_CIPHER_SPEC:
+                        case SslUtils.SSL_CONTENT_TYPE_ALERT:
+                            final int len = SslUtils.getEncryptedPacketLength(in, readerIndex);
+
+                            // Not an SSL/TLS packet
+                            if (len == SslUtils.NOT_ENCRYPTED) {
+                                handshakeFailed = true;
+                                NotSslRecordException e = new NotSslRecordException(
+                                        "not an SSL/TLS record: " + ByteBufUtil.hexDump(in));
+                                in.skipBytes(in.readableBytes());
+
+                                SslUtils.notifyHandshakeFailure(ctx, e);
+                                throw e;
+                            }
+                            if (len == SslUtils.NOT_ENOUGH_DATA ||
+                                    writerIndex - readerIndex - SslUtils.SSL_RECORD_HEADER_LENGTH < len) {
+                                // Not enough data
+                                return;
+                            }
+                            // increase readerIndex and try again.
+                            in.skipBytes(len);
+                            continue;
+                        case SslUtils.SSL_CONTENT_TYPE_HANDSHAKE:
+                            final int majorVersion = in.getUnsignedByte(readerIndex + 1);
+
+                            // SSLv3 or TLS
+                            if (majorVersion == 3) {
+                                final int packetLength = in.getUnsignedShort(readerIndex + 3) +
+                                                         SslUtils.SSL_RECORD_HEADER_LENGTH;
+
+                                if (readableBytes < packetLength) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS134)